registry/event/event_listener.go

```diff
@@ -0,0 +1,193 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package event
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/metadata/service/inmemory"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting"
+	gxset "github.com/dubbogo/gost/container/set"
+	"reflect"
+)
+
+import (
+	"github.com/apache/dubbo-go/common/observer"
+)
+
+// The Service Discovery Changed  Event Listener
+type ServiceInstancesChangedListener struct {
+	registry.ServiceInstancesChangedListenerBase
+	ServiceNames       *gxset.HashSet
+	listeners          map[string]registry.NotifyListener
+	serviceUrls        map[string][]*common.URL
+	revisionToMetadata map[string]*common.MetadataInfo
+	allInstances       map[string][]registry.ServiceInstance
+}
+
+func NewServiceInstancesChangedListener(services *gxset.HashSet) *ServiceInstancesChangedListener {
+	return &ServiceInstancesChangedListener{
+		ServiceNames:       services,
+		listeners:          make(map[string]registry.NotifyListener),
+		serviceUrls:        make(map[string][]*common.URL),
+		revisionToMetadata: make(map[string]*common.MetadataInfo),
+		allInstances:       make(map[string][]registry.ServiceInstance),
+	}
+}
+
+// OnEvent on ServiceInstancesChangedEvent the service instances change event
+func (lstn *ServiceInstancesChangedListener) OnEvent(e observer.Event) error {
+	ce, ok := e.(*registry.ServiceInstancesChangedEvent)
+	if !ok {
+		return nil
+	}
+	var err error
+	lstn.allInstances[ce.ServiceName] = ce.Instances
+	revisionToInstances := make(map[string][]registry.ServiceInstance)
+	newRevisionToMetadata := make(map[string]*common.MetadataInfo)
+	localServiceToRevisions := make(map[*common.ServiceInfo]*gxset.HashSet)
+	protocolRevisionsToUrls := make(map[string]map[*gxset.HashSet][]*common.URL)
+	newServiceURLs := make(map[string][]*common.URL)
+
+	for _, instances := range lstn.allInstances {
+		for _, instance := range instances {
+			revision := instance.GetMetadata()[constant.EXPORTED_SERVICES_REVISION_PROPERTY_NAME]
+			if "0" == revision {
+				logger.Infof("Find instance without valid service metadata: %s", instance.GetHost())
+				continue
+			}
+			subInstances := revisionToInstances[revision]
+			if subInstances == nil {
+				subInstances = make([]registry.ServiceInstance, 8)
+			}
+			revisionToInstances[revision] = append(subInstances, instance)
+			metadataInfo := lstn.revisionToMetadata[revision]
```

Comment:
are u sure the value of revision is not ""? If not, pls check the return value 'ok' of `metadataInfo,ok := lstn.revisionToMetadata[revision]`.

Reply:
I think don't need judge ok